Import Channel and Parcel from amphp/parallel

trowski opened the pull request.

kelunik commented on src/ChannelParser.php:
```diff
@@ -0,0 +1,93 @@
+use Amp\Serialization\Serializer;
+use function Amp\Serialization\encodeUnprintableChars;
+
+final class ChannelParser extends Parser
```
This should likely be internal?

kelunik commented on src/Pipeline/Internal/ConcurrentOperator.php:
```diff
@@ -0,0 +1,165 @@
+ *
+ * @internal
+ */
+final class ConcurrentOperator implements PipelineOperator
```
Can be removed.

kelunik commented on src/Pipeline/functions.php:
```diff
@@ -0,0 +1,25 @@
+function concurrent(Semaphore $semaphore, PipelineOperator ...$operators): PipelineOperator
+{
+    return new Internal\ConcurrentOperator($semaphore, $operators);
+}
```
Can be removed.

kelunik commented on src/PipelineChannel.php:
```diff
@@ -0,0 +1,67 @@
+    public function send(mixed $data): void
+    {
+        if ($data === null) {
+            throw new ChannelException("Cannot send null on a channel");
```
Can we allow this now?

trowski replied:
`null` is currently used to indicate that the channel has closed when returned from `Channel::receive()`.

kelunik commented on src/ChannelledSocket.php:
```diff
@@ -0,0 +1,87 @@
+
+    public function isClosed(): bool
+    {
+        return $this->read->isClosed() || $this->write->isClosed();
```
Should this be &&?

trowski replied:
I was considering a channel to be closed if either end closed, since the API currently doesn't distinguish being able to send or receive.

trowski committed:
Import Channel and Parcel from amphp/parallel

trowski committed:
Remove pipeline operator

trowski committed:
Update for pipeline changes

kelunik commented on src/ChannelledSocket.php:
```diff
@@ -0,0 +1,87 @@
+    }
+
+    /**
+     * @return resource|object|null
```
`object` should probably be removed?

kelunik commented on src/ChannelledSocket.php:
```diff
@@ -0,0 +1,87 @@
+ * @template TSend
+ * @template-implements Channel<TReceive, TSend>
+ */
+final class ChannelledSocket implements Channel, ClosableStream, ResourceStream
```
Why does this class exist instead of using `ChannelledStream` directly?

trowski replied:
Prior, this class had a `close` method where as `ChanneledStream` did not because streams weren't always closable. Since this has changed I don't think this needs to exist.

kelunik commented on src/ChannelledStream.php:
```diff
@@ -0,0 +1,112 @@
+ * @template TSend
+ * @template-implements Channel<TReceive, TSend>
+ */
+final class ChannelledStream implements Channel, ClosableStream
```
This is a `Channel` rather than a `Stream`, so it should probably be named `StreamChannel`?

kelunik commented on src/ConcurrentIteratorChannel.php:
```diff
@@ -0,0 +1,71 @@
+use Amp\Pipeline\Queue;
+
+/**
+ * Creates a channel from a Pipeline and Emitter. The Pipeline emits data to be received on the channel (data
```
Emitter → Queue, Pipeline → ConcurrentIterator

kelunik commented on src/ConcurrentIteratorChannel.php:
```diff
@@ -0,0 +1,71 @@
+ * @template TSend
+ * @template-implements Channel<TReceive, TSend>
+ */
+final class ConcurrentIteratorChannel implements Channel, ClosableStream
```
Should `Channel` simply have a `close` method itself instead of using `ClosableStream`?

trowski replied:
I think that makes more sense, yes. Updated.

kelunik commented on src/ConcurrentIteratorChannel.php:
```diff
@@ -0,0 +1,71 @@
+ * @template TSend
+ * @template-implements Channel<TReceive, TSend>
+ */
+final class ConcurrentIteratorChannel implements Channel, ClosableStream
```
Do we need this class outside of `LocalChannel`? Can this be rolled directly into `LocalChannel`?

trowski replied:
This is used by `createChannelPair()`. I'm not sure how useful `LocalChannel` even is outside of test cases, where the pair could be used instead. Thinking `LocalChannel` should be removed to avoid misuse.

kelunik replied:
`LocalChannel` can be used to synchronize two coroutines with message passing, no?

trowski replied:
No, it can't be. You need to generate a pair to do that.

trowski replied:
Well… unless one coroutine strictly sends and the other strictly receives, but you can also do that with a pair and with less confusion for end users.

kelunik commented on src/SharedMemoryParcel.php:
```diff
@@ -0,0 +1,445 @@
+     * @throws \Error If the size or permissions are invalid.
+     */
+    public static function create(
+        string $id,
```
The `$id` is something we likely should autogenerate, but not sure whether the API of IPC_PRIVATE is exposed.

trowski replied:
Since we need to open the segment in multiple processes, I don't think IPC_PRIVATE applies here.

kelunik commented on src/SharedMemoryParcel.php:
```diff
@@ -0,0 +1,445 @@
+           the old handle.
+        */
+        if (\shmop_size($this->handle) < $size + self::MEM_DATA_OFFSET) {
+            $this->key = $this->key < 0xffffffff ? $this->key + 1 : \random_int(0x10, 0xfffffffe);
```
What's happening here?

trowski replied:
Seems it is picking the next incremental ID or a random ID if the next one is too high? This seems very odd and either we should fail if the memory segment is too small or do something better.

kelunik commented on src/SharedMemoryParcel.php:
```diff
@@ -0,0 +1,445 @@
+     *
+     * @throws ParcelException
+     */
+    private function memOpen(int $key, string $mode, int $permissions, int $size): void
```
`openSegment`?

kelunik commented on src/SharedMemoryParcel.php:
```diff
@@ -0,0 +1,445 @@
+     *
+     * @throws ParcelException
+     */
+    private function memGet(int $offset, int $size): string
```
`readSegment`?

kelunik commented on src/SharedMemoryParcel.php:
```diff
@@ -0,0 +1,445 @@
+     *
+     * @throws ParcelException
+     */
+    private function memSet(int $offset, string $data): void
```
`writeSegment`?

kelunik commented on src/SharedMemoryParcel.php:
```diff
@@ -0,0 +1,445 @@
+     *
+     * @throws ParcelException
+     */
+    private function memDelete(): void
```
`deleteSegment`?

trowski committed:
Channel updates

- Added close() and isClosed() directly to Channel interface.
- Removed ChannelledSocket.
- Renamed ChannelledStream to StreamChannel.

trowski committed:
Update docblock

trowski committed:
Better naming

trowski committed:
Generate queue and shared memory IDs

kelunik commented on src/PosixSemaphore.php:
```diff
@@ -232,21 +228,42 @@ private function open(): void
+        try {
+            do {
+                while (\msg_queue_exists($id = self::$nextId)) {
+                    self::$nextId = self::$nextId % \PHP_INT_MAX + 1;
```
This likely needs to be restricted to 32 bit?

trowski replied:
This worked fine on macOS, but for some systems you're probably right.

kelunik commented on src/SharedMemoryParcel.php:
```diff
@@ -118,29 +117,28 @@ private static function makeKey(string $id): int
     }
 
-    public function getId(): string
+    public function getId(): int
```
Should this be getKey then?

trowski committed:
getID -> getKey

trowski committed:
Limit ID to 32-bit

trowski committed:
Fix styles

trowski committed:
Suppress psalm issues

trowski committed:
How did this line-break end up here?

trowski committed:
Forgot to remove ClosableStream

trowski committed:
Whitelist shmop symbols

trowski committed:
Remove ext-shmop and ext-sysvmsg from composer.json

Prevents testing on Windows.

kelunik commented on src/PosixSemaphore.php:
```diff
@@ -232,21 +229,44 @@ private function open(): void
     {
-        if (\msg_queue_exists($this->key)) {
-            throw new SyncException('A semaphore with that ID already exists');
+        if (!self::$nextId) {
```
Maybe better `self::$nextId === 0` here?

trowski committed:
Delete LocalChannel and make a couple classes internal

trowski committed:
Clearer compare

trowski committed:
Refactor creation error handling

Hard limit of 128 MB for shared memory… but that's huge!

kelunik commented on src/Internal/ChannelParser.php:
```diff
@@ -0,0 +1,94 @@
+     * @param \Closure(mixed):void $closure Closure invoked when data is parsed.
+     * @param Serializer|null $serializer
+     */
+    public function __construct(\Closure $closure, ?Serializer $serializer = null)
```
Maybe name `$closure` → `$onMessage`?

kelunik commented on src/Internal/ChannelParser.php:
```diff
@@ -0,0 +1,94 @@
+     *
+     * @throws ChannelException
+     */
+    public function encode(mixed $data): string
```
Should a `Parser` encode data? :stuck_out_tongue_winking_eye:

trowski replied:
Naming is probably bad… it's encoding what the parser expects. It's internal, so I don't much care.

kelunik replied:
Same ;-)

kelunik commented on src/Internal/ConcurrentIteratorChannel.php:
```diff
@@ -0,0 +1,73 @@
+ * @template TSend
+ * @template-implements Channel<TReceive, TSend>
+ */
+final class ConcurrentIteratorChannel implements Channel
```
Misses `@internal`.

kelunik commented on src/SharedMemoryParcel.php:
```diff
@@ -0,0 +1,481 @@
+        }
+
+        // Invalidate the memory block by setting its state to FREED.
+        $this->writeSegment(self::generateHeader(self::STATE_FREED, 0, 0));
+
+        // Request the block to be deleted, then close our local handle.
+        $this->deleteSegment();
+        $this->handle = null;
```
So the original creator is always the real owner of the segment?

trowski replied:
AFAIK as long as any process has a handle to the segment the OS will keep it around. So there isn't an 'owner' per-say.

kelunik replied:
Yeah, but the original owner will write the `STATE_FREED` header, no?

trowski replied:
Yes, you're right. Not sure why there's an explicit call to unset mutex below either…

kelunik commented on src/StreamChannel.php:
```diff
@@ -0,0 +1,112 @@
+
+                $parser->push($chunk);
+
+                while (!$received->isEmpty()) {
+                    yield $received->shift();
+                }
```
Should we use `Queue` here, so we can allow a configurable buffer size?

trowski replied:
Let's leave it for now and revisit this some other time.

kelunik commented on src/functions.php:
```diff
@@ -25,3 +28,20 @@ function synchronized(Semaphore $semaphore, \Closure $synchronized, mixed ...$ar
 }
+
+/**
+ * @template TReceive
+ * @template TSend
```
These can't be inferred, so should we add pseudo class-string arguments to infer the types?

trowski replied:
I just dropped the template types on the function since they aren't useful. Channels will have to be annotated if someone actually cares to limit it anyway.

trowski committed:
Allow sending null on channel

trowski committed:
$closure -> $onMessage

kelunik commented on src/Channel.php:
```diff
@@ -0,0 +1,43 @@
+     * @param Cancellation|null $cancellation Cancels waiting for the next value. Note the next value is not discarded
+     * if the operation is cancelled, rather it will be returned from the next call to this method.
+     *
+     * @return TReceive|null Data received or null if the channel closed.
```
```suggestion
     * @return TReceive Data received.
```

kelunik commented on src/Channel.php:
```diff
@@ -0,0 +1,43 @@
+     *
+     * @return TReceive|null Data received or null if the channel closed.
+     *
+     * @throws ChannelException If receiving from the channel fails.
```
```suggestion
     * @throws ChannelException If receiving from the channel fails or the channel closed.
```

kelunik commented on src/Channel.php:
```diff
@@ -0,0 +1,43 @@
+    /**
+     * @param TSend $data
+     *
+     * @throws ChannelException If sending on the channel fails.
```
```suggestion
     * @throws ChannelException If sending on the channel fails or the channel is already closed.
```

trowski committed:
Update src/Channel.php

Co-authored-by: Niklas Keller <me@kelunik.com>

trowski committed:
Update src/Channel.php

Co-authored-by: Niklas Keller <me@kelunik.com>

trowski committed:
Update src/Channel.php

Co-authored-by: Niklas Keller <me@kelunik.com>

kelunik commented on src/LocalParcel.php:
```diff
@@ -0,0 +1,38 @@
+     * @param T $value
+     */
+    public function __construct(
+        private Semaphore $semaphore,
```
It should require a `Mutex` instead, no?

trowski replied:
We allow a `Semaphore` on `Amp\Sync\synchronized()`. I would generally think both should be a `Mutex` actually.

kelunik replied:
Yeah, `synchronized` should be a `Mutex`, too.

kelunik commented on src/PosixSemaphore.php:
```diff
@@ -15,37 +15,39 @@
 final class PosixSemaphore implements Semaphore
 {
     private const LATENCY_TIMEOUT = 0.01;
+    private const MAX_ID = 0xffffffff;
```
We should probably use the signed int maximum instead of the unsigned int maxium?

trowski replied:
Not sure it matters, any non-zero value works locally.

kelunik replied:
Let's see whether it'll be an issue.

kelunik commented on src/SharedMemoryParcel.php:
```diff
@@ -0,0 +1,481 @@
+        return $parcel;
+    }
+
+    private static function makeKey(string $id): int
+    {
+        return (int) \abs(\unpack("l", \md5($id, true))[1]);
+    }
```
This can be dropped, no?

kelunik commented on src/SharedMemoryParcel.php:
```diff
@@ -0,0 +1,481 @@
+    private Mutex $mutex;
+
+    /** @var \Shmop An open handle to the shared memory segment. */
+    private ?\Shmop $handle = null;
```
On which version has it been converted to an object?

trowski replied:
8.0

kelunik commented on src/StreamChannel.php:
```diff
@@ -0,0 +1,112 @@
+    public function receive(?Cancellation $cancellation = null): mixed
+    {
+        if (!$this->iterator->continue($cancellation)) {
+            return null;
```
This needs to throw now.

trowski replied:
Technically this won't happen now, but yeah, in case it does.

trowski committed:
Missed null return

trowski committed:
Remove unused method

trowski committed:
Limit ID to 2^31

trowski committed:
Mutex instead of semaphore

trowski committed:
Skip style fixer on windows

trowski committed:
Remove freed state
